### PR TITLE
Add validation for failover

### PR DESCRIPTION
**Pull request by Poor12**
**What type of PR is this?**
/kind api-change
/kind cleanup

**What this PR does / why we need it**:
Add validation for Failover. `TolerationSeconds`, `DelaySeconds` and `BlockPredecessorSeconds` cannot < 0.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

### CODE REVIEW

**Review by RainbowMango — COMMENTED** [review]
/assign

File: pkg/util/validation/validation.go [hunk 1]
```diff
@@ -142,6 +143,42 @@ func ValidateSpreadConstraint(spreadConstraints []policyv1alpha1.SpreadConstrain
...
+		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.BlockPredecessorSeconds, "the blockPredecessorSeconds is less than zero"))
+	}
+
+	if *applicationFailoverBehavior.DecisionConditions.TolerationSeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.DecisionConditions.TolerationSeconds, "the tolerationSeconds is less than zero"))
+	}
+
+	preConditions := applicationFailoverBehavior.PreConditions
+	if preConditions != nil && preConditions.DelaySeconds != nil && *preConditions.DelaySeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, *preConditions.DelaySeconds, "the delaySeconds is less than zero"))
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:176]:
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath, *preConditions.DelaySeconds, "must be greater than or equal to 0"))
```

File: pkg/util/validation/validation.go [hunk 2]
```diff
@@ -142,6 +143,42 @@ func ValidateSpreadConstraint(spreadConstraints []policyv1alpha1.SpreadConstrain
...
+	if applicationFailoverBehavior == nil {
+		return nil
+	}
+
+	if *applicationFailoverBehavior.BlockPredecessorSeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.BlockPredecessorSeconds, "the blockPredecessorSeconds is less than zero"))
+	}
+
+	if *applicationFailoverBehavior.DecisionConditions.TolerationSeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.DecisionConditions.TolerationSeconds, "the tolerationSeconds is less than zero"))
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:171]:
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.DecisionConditions.TolerationSeconds, "must be greater than or equal to 0"))
```

File: pkg/util/validation/validation.go [hunk 3]
```diff
@@ -142,6 +143,42 @@ func ValidateSpreadConstraint(spreadConstraints []policyv1alpha1.SpreadConstrain
...
+// ValidateApplicationFailover validates that the application failover is correctly defined.
+func ValidateApplicationFailover(applicationFailoverBehavior *policyv1alpha1.ApplicationFailoverBehavior, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if applicationFailoverBehavior == nil {
+		return nil
+	}
+
+	if *applicationFailoverBehavior.BlockPredecessorSeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, *applicationFailoverBehavior.BlockPredecessorSeconds, "the blockPredecessorSeconds is less than zero"))
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:167]: Same as below.

Comment (RainbowMango) [reply pkg/util/validation/validation.go:167]: It'd be great to validate the fields by the order they appear in the struct.

File: pkg/util/validation/validation_test.go [hunk 4]
```diff
@@ -554,3 +555,76 @@ func TestValidatePropagationSpec(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateApplicationFailover(t *testing.T) {
+	var correctSeconds int32 = 100
```
Comment (Poor12) [inline pkg/util/validation/validation_test.go:560]: Consider all Seconds are point type, we may still need these temporary variables.

Comment (RainbowMango) [reply pkg/util/validation/validation_test.go:560]: pointer.Int32(100)

File: pkg/util/validation/validation_test.go [hunk 5]
```diff
@@ -554,3 +555,76 @@ func TestValidatePropagationSpec(t *testing.T) {
...
+		},
+		{
+			name: "the delaySeconds is less than zero",
+			applicationFailoverBehavior: &policyv1alpha1.ApplicationFailoverBehavior{
+				BlockPredecessorSeconds: &correctSeconds,
+				DecisionConditions: policyv1alpha1.DecisionConditions{
+					TolerationSeconds: &correctSeconds,
+				},
+				PreConditions: &policyv1alpha1.PreConditions{
+					DelaySeconds: &errSeconds,
```
Comment (RainbowMango) [inline pkg/util/validation/validation_test.go:580]:
```suggestion
					DelaySeconds: pointer.Int32(-1),
```

File: pkg/util/validation/validation_test.go (hunk 4 above)
Comment (Poor12) [reply pkg/util/validation/validation_test.go]: Thanks for reminding.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**